WIP some more consensus (#8)
cardano-scaling/cardano-blueprint: src/consensus/consensus.md

# The Consensus Layer

> ```mermaid

Author: How so? it does render for me even working on Windows. Do you have `mdbook-mermaid` installed?

> ```mermaid

Reviewer: ah that's the point :) I don't use nix so 🤷

> ```mermaid

Author: Then you could use `cargo install mdbook mdbook-katex mdbook-mermaid mdbook-alerts mdbook-toc` and it should probably work fine. I don't use nix either.

> ```mermaid

Reviewer: Who reads `README`s 😂 ? Thanks for the tip :)

flowchart TD
    A("Consensus") --> C("Select From Candidate Chains")
    C("Select From Candidate Chains") --> A1("Fetch Chains from Peers")
    C("Select From Candidate Chains") --> A2("Ask Ledger to Validate Blocks")

    A("Consensus") --> E("Store and Serve Historical Chain")

    A("Consensus") --> D("Accept transactions into Mempool")
    D("Accept transactions into Mempool") --> B1("Ask Ledger to Validate Transations")
    D("Accept transactions into Mempool") --> B2("Fetch Transactions from Peers")

    A("Consensus") --> B("Mint Blocks")

    B("Mint Blocks") --> B1("Ask Ledger to Validate Transations")
```

> The Consensus layer is responsible for choosing among the different chains that might co-exist on the network. The candidate chains may arise both from honest and adversarial participation. The Consensus Layer must implement the consensus protocol of the current Cardano era to know what is the right candidate chain to choose.

Author: Yeah, I have removed this in favor of the later paragraphs, with links to the appropriate sections

> To replicate the blockchain across all Cardano nodes, the Consensus layer must distribute the selected chain to its peers.

Author: This paragraph is gone, but I added the suggestion somewhere else.

The Consensus layer is also responsible for producing new blocks according to the current consensus protocol. For the new blocks to include meaningful data, Consensus has to receive and maintain a collection of submitted transactions that are not in the chain yet.

<!-- toc -->

## The Consensus Protocol in Cardano

> The consensus protocol has three primary responsibilities:

Author: No, the responsibilities belong to different parts of the system. The Protocol responsibilities are more limited than the ones of Consensus.

- Chain selection, and chain propagation: Competing chains arise when two or more nodes
  extend the chain with different blocks. This can happen when nodes are not aware of
  each other’s blocks due to temporarily network delays or partitioning, but depending on
  the particular choice of consensus algorithm it can also happen in the normal course of
  events. When it happens, it is the responsibility of the consensus protocol to choose
  between these competing chains.

  The process of selecting chains decides which block bodies to download and in turn which
  chains are diffused downstream over the network. Albeit the fact that Consensus is
  theoretically ignorant on how chains arrive at its door, its decisions are critical on
  the overall flow of data of the system.

> - Transaction validation and diffusion: Consensus is in charge of keeping a record of data

Author: Yes, good point. I see now that I mixed this up.

> - Transaction validation and diffusion: Consensus is in charge of keeping a record of data

Author: I think it is clearer now that I separated the protocol and the component responsibilities

| Mary                                       | Ouroboros TPraos                      |                                                                                                                                    |
| Alonzo                                     | Ouroboros TPraos                      |                                                                                                                                    |
| Babbage                                    | Ouroboros Praos                       | [Paper](https://iohk.io/en/research/library/papers/ouroboros-praos-an-adaptively-secure-semi-synchronous-proof-of-stake-protocol/) |
| Conway                                     | Ouroboros Praos                       |                                                                                                                                    |

Each of these protocols defines how to fulfill the responsibilities above. Regarding
validity of blocks, the Consensus layer can remain oblivious to the details on validation
and rely on the Ledger layer to make such judgment, based on the particular era in effect.

## Selecting chains

> _Chain selection_ is the mechanism, specified in the Protocol associated to the current

Author: It is now defined in a single place.

era in effect in the network, through which nodes acquire and diffuse the best chain in

> the network. It is important to note that **all nodes** participating in the network are in

Reviewer: Perhaps would be interesting to outline that this implies even non block forging node contribute to the security of the network?

Author: Done :+1:

- If blocks can be almost fully validated in constant time based on looking at only a small fixed
size block header, then honest nodes can validate candidate chains with a small bounded
amount of work.

- It also enables a design where a node can see blocks available from many immediate
peers but can choose to download each block body of interest just once (from a peer of its
choosing from which it is available). This saves network bandwidth.

In the case of Ouroboros, all the cryptographic consensus evidence is packed into the
block header, leaving the block body containing only the ledger data, and check that the
block has been signed by a node that is the slot leader. If we validate this in the
context of a chain of headers, then we can establish this is a plausible candidate
chain, thus we eliminate several potential resource draining attacks.

So the design at this stage involves transmitting chains of headers rather than whole
blocks, and using a secondary mechanism to download block bodies of interest.

The Consensus chain selection is involved with chains of block headers, whereas the
validity check is performed by the Ledger rules, effectively separating concerns.

> > [!NOTE]

Author: I think it is fine now that I reworded it

> Chain selection is performed based on data from the previous headers, so there is some state
> (specific to each protocol) that influences the selection, which we will call _chain
> state_, as opposed to the state of the ledger, which we will call _ledger state_.

### The `k` security parameter

Both Ouroboros Classic and Ouroboros Praos are based on a chain
selection rule that imposes a maximum rollback condition: alternative
chains to a node's current chain that fork off more than a certain
number of blocks ago are never considered for adoption. This limit is
known as the security parameter, and is usually denoted by `k`, which

> on mainnet is set to 2160 blocks. The analysis shows that nodes will

Reviewer: ```suggestion
on mainnet is currently set to 2160 blocks. Analysis from research has shown this parameter ensures nodes will 
```

> not diverge more than `k` blocks under honest circumstances.

Reviewer: ```suggestion
not diverge more than `k` blocks under honest circumstances with sufficiently high probability.
```

> Ouroboros BFT does not impose a maximum rollback, but adding such a

Reviewer: Do we care about BFT at this stage?

Author: I think we care about it at least for implementation reasons, like replaying the chain needs to know how to process blocks from BFT.

requirement does not change the protocol in any fundamental way.

In order to switch to an alternative chain, the Consensus layer needs to evaluate the
validity of such a chain. In order to evaluate the validity of a particular header, we
need to have a Chain State at the predecessor header. In order to evaluate the validity of
a particular block, we need to have a Ledger State at the predecessor block. As applying
blocks and applying headers are not invertible operations, this is usually solved by
maintaining the last `k+1` ledger states and chain states in memory, the alternative being
replaying again the chain since Genesis which is infeasible.

### Chain selection during syncing (Ouroboros Genesis)

When a node is syncing with the network, there is little benefit in having it
participate in producing blocks or validating transactions as it doesn't have a
ledger state close enough to the tip of the chain. This in particularly means
that all the responsibilities of a node are discharged from such syncing node

> except the duty to select the best chain it can.

Author: Maybe, but I don't think it will have a real impact.

> except the duty to select the best chain it can.

Author: Even further, I think syncing nodes _do_ run the forging loop and transaction validation.

> except the duty to select the best chain it can.

Author: Also this paragraph is gone now

On BFT/Classic, as only seven authorized Core nodes were able to make new
blocks, there can be no mistake on the chosen chain (assuming Core
nodes keys aren't compromised and used to create an old adversarial
chain in the Byron era).

However, in Praos the rule for selecting the best chain among two competing ones
is fundamentally based on length of the chain. The protocol assumes
instantaneous transmission of whole chains, which would mean that the perceived
length of a candidate would indeed be the real length. For fundamental reasons,
sending such amount of data instantaneously is not possible and instead the data
is streamed to the syncing node. This has the caveat of nodes only knowing the
length of the received prefix of the candidate. This exposed the risk of
adversaries tricking nodes into committing to an adversarial chain when syncing

> with the network.

Author: Of course, using Praos that is what they would do.

> To mitigate this risk in Praos, the IOG researchers published [Ouroboros

Author: The Genesis chain selection section specifies this now

Genesis](https://iohk.io/en/research/library/papers/ouroboros-genesis-composable-proof-of-stake-blockchains-with-dynamic-availability/)
which, in simple terms, selects the best candidate based on the density of the
chain at the intersection between the candidates, leveraging the property of
Ouroboros that the honest chain will have more blocks than any other adversarial
chain in a specific window of slots from the intersection point.

Once the node finishes syncing with the network, this rule gracefully converges
into the usual length-based comparison used in Praos, so the node can safely
switch to running only Ouroboros Praos.

### Chain diffusion MiniProtocols

> To fulfill the requirements above, the Consensus layer implements `ChainSync` and

Author: Yes, these are the ChainSync and BlockFetch miniprotocols precisely.

> To fulfill the requirements above, the Consensus layer implements `ChainSync` and

Author: Also this part is gone now

`BlockFetch`, miniprotocols to diffuse the blockchain through the network:

| MiniProtocol | Purpose                                                                  |
|:-------------|:-------------------------------------------------------------------------|
| `ChainSync`  | Synchronize the changes on the nodes tips and transmit chains of headers |
| `BlockFetch` | Transmit specific blocks                                                 |

The specifications of the state machines of these miniprotocols are described in the
Network documentation
([design](https://ouroboros-network.cardano.intersectmbo.org/pdfs/network-design/network-design.pdf)
and
[spec](https://ouroboros-network.cardano.intersectmbo.org/pdfs/network-spec/network-spec.pdf)).

> ## Transaction bookkeeping

Reviewer: ```suggestion
## Transactions Mempool
```

Author: I moved this to a separate section, called "Mempool"

For blocks to have useful data, they have to contain transactions, which are codifications
of operations on the Ledger state that only some authorized actors can enact. Notice that
such transactions are what makes up the contents of the block. In one way, one could see
the ledger as being able to validate transactions, and by implication full blocks as those

> are mainly collections of transactions.

Author: I think it is useful to write this down here. A user knows what transactions are but maybe have not thought about what transactions are for Consensus and how the ledger interacts with them on behalf of consensus

Note that although the nodes that will _use_ these transactions to create blocks are only
the ones that forge new blocks, other nodes are mainly used to diffuse transactions to the
majority of the network. It is expected that **all nodes** that participate in the network
are able to receive, validate and distribute transactions.

> The Mempool is the abstract component that stores transactions which are valid on top of

Author: I think I am precisely mentioning that in:

> The Mempool is the abstract component that stores transactions which are valid on top of the latest known Ledger State.

and 

> - Re-validation on a new selected chain: it is useless to keep transactions that are no
  longer valid in the mempool.

> The Mempool is the abstract component that stores transactions which are valid on top of

Author: Yes, let's go deeper in the mempool stuff in some other PR.

> the latest known Ledger State. Being valid implies in particular that the transaction is

Author: Well, technically it is the latest known Ledger State in the whole system, and when that one changes in the LedgerDB it should instantly change in the mempool, so I wouldn't make this distinction.

pending, i.e. it has not been included in any prior block ever, as in Cardano every valid
transaction must consume at least one UTxO.

Usually pending transactions are discarded when the node is shut down, however this is not
a hard requirement. It is conceivable for a node to store the received pending
transactions on the disk and re-evaluate them on startup. As IOG's Haskell implementation
does not store the pending transactions on the disk, the mempool is currently considered
part of the Consensus layer, and not part of the Storage layer, as it does not store data
on the disk.

> The `TxSubmission` miniprotocol is the one used to diffuse the pending transactions

Reviewer: Perhaps nitpicking but is it not `TxSubmission2`?

Author: I think so, good point.

Author: The `2` is probably a thing for Network to discuss.

And a `pending transaction` is a transaction that is valid on the latest ledger state. If a transaction is pending it is in particular valid.

through the network. Its state machine specification and general mechanics are described
in the [Networking
spec](https://ouroboros-network.cardano.intersectmbo.org/pdfs/network-spec/network-spec.pdf).

> Any mempool implementation has some requirements to allow diffusing transactions via

Author: There is now a page for TxSubmission

`TxSubmission` and to fulfill the requirements of the Consensus layer in terms of forging
blocks:
- Acquiring a snapshot of valid transactions for a newly forged block should be as fast as
  possible, as it will delay all other steps in the diffusion of such a block.
- Index-based access to pending transactions, to fulfill the `TxSubmission` protocol requirements.
- Re-validation on a new selected chain: it is useless to keep transactions that are no
  longer valid in the mempool.

## Forging new blocks

Forging new blocks is the process of packing data that has not yet been included in the
blockchain (pending transactions) into a fresh block. Blocks are signed by stake pool

> credentials, so only the nodes which are configured with such credentials strictly need to

Author: Those are the nodes that strictly speaking need a forging logic. Non-block producers could run without such logic implemented.

implement the forging mechanism.

### The leadership check

Every slot, the stake pool (whose keys the node was configured with) might be entitled to
produce a block, depending on the rules of the particular Protocol that is running at the
tip of the Cardano chain. For Praos, slots have a duration of one second. The probability
of a node being elected in a slot is controlled via the _active slot coefficient_
parameter (commonly referred to as `f`) which on mainnet is 0.05. This means that a block
is expected once every 20 slots on average.

In Ouroboros, a block consists of a sequence of transactions, so forging a block would
look something similar to the following sequence of actions:

> 1. Determine wheter my stake pool is entitled to create a block on this slot.

Reviewer: ```suggestion
1. Determine whether my stake pool is entitled to create a block on this slot.
```

> 2. Acquire a list of valid transactions with respect to the best Ledger State my node knows.

Reviewer: Perhaps worth mentioning transactions should be selected w.r.t limits of block size, execution units

The newly created block is then fed into the chain selection logic, which expectedly will
select it and then diffuse it via the usual mechanisms described above.

### Multi-leader slots

There exists the possibility of multiple stake pools being elected in the same slot. If
the nodes of both pools produce blocks, a momentary fork will exist in the chain which we
call _a slot battle_. It is guaranteed by Ouroboros Praos that only one of those blocks
will end up in the honest chain, and that such fork will not survive more than the time it
will take for either of the candidates to grow to `k` blocks.

Two stake pools might be elected in very close slots, which implies that if the block of
the first stake pool doesn't arrive the node of the second stake pool on time, it will not
be considered when the second node forges its block, creating again a short-lived fork,
which we call _height battles_. In order to avoid this situation it is of utter importance
that the forging of a block and its diffusion through the network (first its header via
ChainSync, then its body via BlockFetch) is as fast as possible.

## MiniProtocols

> The miniprotocols mentioned above (`ChainSync`, `BlockFetch` and `TxSubmission`) are one

Author: I am talking here about the ones I know and mentioning that they are not the only option. I think it is a fair point.

of the endless possibility of mechanisms used for data difussion. The Networking design
document has many more insights on why these protocols were implemented, and how they

> differ from other off-the-shelve mechanisms.

Reviewer: ```suggestion
differ from other off-the-shelf mechanisms.
```

Although it is conceivable having other alternative mechanisms to exchange the data, these
miniprotocols are for now the common language spoken by the nodes in the Cardano network,
and as such it is expected that all node implementations use them, or at least are
_capable of using them_ to communicate with the rest of the network.

> ## Resilience of the Consensus layer

Author: I have no proof that this principle is followed in Network/Ledger. I'd rather keep it in the Consensus section.
